### ISSUE TITLE
TLS connection fails

### ISSUE THREAD

**Issue by dedani** [SergioBenitez/Rocket#650]
Hi, I try to setup a shopping list site (based on the 'todo' example) with Rocket for personal use hosted on a raspberry. Without the 'tls' feature everything works just fine. With 'tls' things don't work any more. Specifically, after Firefox asked be about adding the certification loading the page failed.

I generate the certificate with:
```
# openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout key.pem -out cert.pem
I enter the IP of the server when asked (192.168.178.200).
```

The server is started with:
```
# sudo DATABASE_URL=db/db.sql ROCKET_ENV=stage ROCKET_ADDRESS=192.168.178.200 ROCKET_PORT=443 ./target/debug/postiliste
🔧  Configured for staging.
    => address: 192.168.178.200
    => port: 443
    => log: normal
    => workers: 8
    => secret key: generated
    => limits: forms = 32KiB
    => tls: enabled
    => [extra] template_dir: "static"
    => [extra] database_url: "db/db.sql"
🛰  Mounting '/':
    => GET /
    => GET /<path..> [5]
🛰  Mounting '/item/':
    => POST /item
    => PUT /item/<id>
🚀  Rocket has launched from https://192.168.178.200:443
```

curl (I found this on the internet. I guess it helps):
```
# curl -v -L --cacert cert.pem https://192.168.178.200
* Rebuilt URL to: https://192.168.178.200/
*   Trying 192.168.178.200...
* TCP_NODELAY set
* Connected to 192.168.178.200 (192.168.178.200) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
  CAfile: cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to 192.168.178.200:443 
* stopped the pause stream!
* Closing connection 0
curl: (35) OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to 192.168.178.200:443 
```

I get the exact same behaviour and messages when I try this with the 'tls' example.
Is this a bug or am I doing something wrong?

(This is my first web application, so go easy on me.)

**Comment by SergioBenitez** [comment]
I can't replicate this. Please mention:

1. The version of Rocket you're using. Ensure it's the latest, if possible.

2. The operating system (distribution and version) where the issue occurs.

**Comment by dedani** [comment]
I'm using 0.3.11, same as in the 'tls' example which is also not working. It's probably easier to diagnose the example.

I'm having three systems all running opensuse tumbleweed, fully updated -a desktop, a notebook and a raspberry pi 3 B. In the end I want to run the server on the raspberry.

I can run the 'tls' example and my page on the desktop and notebook. Everything works as expected there. But if run on the Pi, I can't connect. If I try to go to the site with firefox, it asks me to get and allow the certificate. But after that I just get a "Secure Connection Failed" site. Similar with Chromium.

**Comment by dedani** [comment]
I tried to run Rocket on Raspbian and it worked. Seams like it's a problem of opensuse or a package with what they ship.

**Comment by SergioBenitez** [comment]
This is interesting. Perhaps @ctz has something to say about this.

**Comment by ctz** [comment]
This could possibly be https://github.com/briansmith/ring/issues/628 ?

@dedani if you still have the Raspberry Pi in the same configuration where it reproduces the problem, can you do `cargo test` in the ring crate?

**Comment by dedani** [comment]
I run cargo test both Tumbleweed and Raspbian:
on the Raspbian, all tests passes:
```
> uname -m -r
4.14.34-v7+ armv7l
> rustup show   
# ...
active toolchain
----------------
nightly-armv7-unknown-linux-gnueabihf (default)
rustc 1.28.0-nightly (4ecf12bf0 2018-06-02)
> git clone https://github.com/briansmith/ring
# ...
> cargo test
# ...
test result: ok. 12 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out
```

on Tumbleweed it fails:
```
> uname -m -r
4.16.6-1-default aarch64
> rustup show   
# ...
active toolchain
----------------
nightly-aarch64-unknown-linux-gnu (default)
rustc 1.28.0-nightly (4ecf12bf0 2018-06-02)
git clone https://github.com/briansmith/ring
# ...
cargo test
# ...
thread 'aead_aes_gcm_128' panicked at 'Test failed.', src/test.rs:352:9
failures:
    aead_aes_gcm_128
    aead_aes_gcm_256
test result: FAILED. 1 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out
error: test failed, to rerun pass '--test aead_tests'
```
[ring_test_tumbleweed.log](https://github.com/SergioBenitez/Rocket/files/2065963/ring_test_tumbleweed.log)

**Comment by briansmith** [comment]
> I run cargo test both Tumbleweed and Raspbian. Also they have a different architecture: arm7l vs aarch64.

Please see https://github.com/briansmith/ring/issues/628#issuecomment-455726680. I believe this is fixed in *ring* 0.14.2.

**Comment by SergioBenitez** [comment]
This appears to have been resolved upstream. Closing out.
